pages/en/snapps/how-to-write-a-snapp.mdx

```diff
@@ -0,0 +1,387 @@
+```
+
+You are free to add additional parameters, like the initial values of your state.
+
+All state variables have to be initialized and wrapped with `State.init()`.
+
+##### Methods and state updates
+
+To update your snapp state and trigger token transfers, a smart contract needs custom methods.
+SnarkyJS allows you to call one or more methods within one transaction.
+
+You declare methods using the `@method` decorator:
+
+```ts
+class HelloWorld extends SmartContract {
+  @state(Field) x: State<Field>;
+
+  // ...
+
+  @method async increment() {
+    const x = await this.x.get();
+    console.log("Current state", x);
+    this.x.set(x.add(1));
+  }
+}
+```
+
+The `increment()` method in our example is async because it calls `this.x.get()`.
+This is how you can fetch the current state of a snapp account from the chain.
+
+The method then logs the current state, and sets the new state to `x + 1` using `this.x.set()`.
+
+The methods `.get()` and `.set()` are defined on every state property.
+
+##### Assertions
+
+Let's modify the `increment()` method to take in a parameter:
+
+```ts
+class HelloWorld extends SmartContract {
+  @state(Field) x: State<Field>;
+
+  // ...
+
+  @method async increment(xPlus1: Field) {
+    const x = await this.x.get();
+    x.add(1).assertEquals(xPlus1);
+    this.x.set(xPlus1);
+  }
+}
+```
+
+Here, after obtaining the current state `x`, we make an _assertion_: `x.add(1).assertEquals(xPlus1);`.
+
+If the assertion fails, SnarkyJS will throw an error and not submit the transaction containing the method call.
+On the other hand, if it succeeds, it will become a part of the proof that is sent to the chain and that is verified on-chain.
+
+Because of this, our new version of `increment()` is guaranteed to behave like the previous version: It can only ever
+update the state `x` to `x + 1`.
+
+Assertions can be incredibly useful to constrain state updates. Common assertions you may use are:
+
+<!-- prettier-ignore -->
+```ts
+x.assertEquals(y); // x = y
+x.assertBoolean(); // x = 0 or x = 1
+x.assertLt(y);     // x < y
+x.assertLte(y);    // x <= y
+x.assertGt(y);     // x > y
+x.assertGte(y);    // x >= y
+```
+
+For a full list, see the <DocLink copy="SnarkyJS reference" url="/snapps/snarkyjs-reference" />.
+
+##### Public and private inputs
+
+While the state of a snapp is **public**, method parameters are **private**.
+
+When a smart contract method is called, its code will execute on the client and generate a proof, which is then sent to the network to update the chain.
+The proof uses zero-knowledge to hide inputs and details of the computation
+.
+
+<!-- The proof has _public inputs_ that block producers need to see -- for example, any state updates triggered by our method. -->
+
+The only way method parameters can be exposed is when the computation explicitly exposes them, as
+in our last example where the input was directly stored in the public state: `this.x.set(xPlus1);`
```

Comment:
Does it make sense to add a sentence at the end?: `Otherwise, parameters will remain private and will never be seen by the Mina network.`